feat(self-hosted): convert experimental env vars to config options (#29154)
renovatebot/renovate: docs/usage/self-hosted-configuration.md

         "customEnvVariables": {"SECRET_TOKEN": "{{ secrets.SECRET_TOKEN }}"},
```

## deleteConfigFile

> If set to `true` Renovate tries to delete the self-hosted config file after reading it.
> You can set the config file Renovate should read with the `RENOVATE_CONFIG_FILE` environment variable.
> 
> The process that runs Renovate must have the correct permissions to delete the config file.

Reviewer: ```suggestion
If set to `true` Renovate tries to delete the self-hosted config file after reading it.

The process that runs Renovate must have the correct permissions to delete the config file.

<!-- prettier-ignore -->
!!! tip
    You can tell Renovate where to find your config file with the `RENOVATE_CONFIG_FILE` environment variable.
```

I think we should:

- first explain how the `deleteConfigFile` works
- then give the tip about the `RENOVATE_CONFIG_FILE` environment variable

So I've moved the paragraphs around, and added a new tip admonition.

Information provided mainly in debug log level.

## eagerGlobalExtends

> Resolve and merge `globalExtends` presets before other global config, instead of after.

Reviewer: ```suggestion
By default, Renovate resolves and merges `globalExtends` presets _after_ other global config.

But if you set `eagerGlobalExtends` to ..., then Renovate will resolve and merge `globalExtends` presets _before_ the global config.
```

Before applying my suggestion, please:

- [ ] Double check if the sentences have the correct logic!
- [ ] Replace the `...` bit with the correct instruction

    When Renovate resolves `globalExtends` it does not fully process the configuration.
    This means that Renovate does not have the authentication it needs to fetch private things.

## globalExtendsEager

By default, Renovate resolves and merges `globalExtends` presets _after_ other global config.

But if you set `globalExtendsEager` to `true`, then Renovate will resolve and merge `globalExtends` presets _before_ the global config.

> 

Reviewer 2: I just realized that we should reverse this default and drop this option in v38.

Maybe better to remove this from the migration?

!!! warning
    When you set `httpCacheTtlDays` to `0`, Renovate will remove the cached HTTP data.

## ignoreNodeWarn

Using this option, you can suppress the default warning when a deprecated version of Node.js is used to run Renovate.

> By default, Renovate logs a WARN debug message if this option is set to true, Renovate will log an INFO debug message instead.

Reviewer: ```suggestion
By default, Renovate logs a `WARN` debug message if this option is set to true, Renovate will log an `INFO` debug message instead.
```

!!! warning
    When you set `httpCacheTtlDays` to `0`, Renovate will remove the cached HTTP data.

## ignoreNodeWarn

Using this option, you can suppress the default warning when a deprecated version of Node.js is used to run Renovate.

By default, Renovate logs a `WARN` debug message if this option is set to true, Renovate will log an `INFO` debug message instead.

> 

Reviewer 2: This should be removed too, because how logLevelRemap can be used instead

Author: https://github.com/renovatebot/renovate/pull/29835

| `requireConfig=optional` | An onboarding PR will be created if no config file exists. If the onboarding PR is closed and there's no config file, the repository will be processed. | Repository is processed regardless of config file presence.   |
| `requireConfig=ignored`  | No onboarding PR will be created and repo will be processed while ignoring any config file present.                                                     | Repository is processed, any config file is ignored.          |

## s3Endpoint

> If set, Renovate will use this string as the `endpoint` when instantiating the AWS S3 client.

Reviewer: Is there a simpler word than `instantiate` that we can use here?

Maybe `when starting the AWS S3 client`?

## s3PathStyle

> If set, Renovate will enable `forcePathStyle` when instantiating the AWS S3 client.

Reviewer: Same comment about instantiate here.

> > Whether to force path style URLs for S3 objects (e.g., `https://s3.amazonaws.com//` instead of `https://.s3.amazonaws.com/`)

Reviewer: ```suggestion
> Whether to force path-style URLs for S3 objects (e.g., `https://s3.amazonaws.com//` instead of `https://.s3.amazonaws.com/`)
```

It feels like we need ` -` here, but this is a direct quote, I think?

| `requireConfig=optional` | An onboarding PR will be created if no config file exists. If the onboarding PR is closed and there's no config file, the repository will be processed. | Repository is processed regardless of config file presence.   |
| `requireConfig=ignored`  | No onboarding PR will be created and repo will be processed while ignoring any config file present.                                                     | Repository is processed, any config file is ignored.          |

## s3Endpoint

If set, Renovate will use this string as the `endpoint` when creating the AWS S3 client instance.

## s3PathStyle

> If set, Renovate will enable `forcePathStyle` when creating the AWS S3 client instance.
> 
> > Whether to force path-style URLs for S3 objects (e.g., `https://s3.amazonaws.com//` instead of `https://.s3.amazonaws.com/`)
> 
> Source: [AWS S3 documentation - Interface BucketEndpointInputConfig](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/interfaces/bucketendpointinputconfig.html)

Reviewer: ```suggestion
If set, Renovate will enable `forcePathStyle` when creating the AWS S3 client instance.

Quote from [AWS S3 documentation - Interface BucketEndpointInputConfig](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/interfaces/bucketendpointinputconfig.html):

> Whether to force path-style URLs for S3 objects (e.g., `https://s3.amazonaws.com//` instead of `https://.s3.amazonaws.com/`)
```

Switching around the order, with a small rewrite.

> > Whether to force path-style URLs for S3 objects (e.g., `https://s3.amazonaws.com//` instead of `https://.s3.amazonaws.com/`)

Reviewer 3: ```suggestion
> Whether to force path-style URLs for S3 objects (e.g., `https://s3.amazonaws.com/bucket/` instead of `https://bucket.s3.amazonaws.com/`)
```

i think this is a more clear sample

> Quote from [AWS S3 documentation - Interface BucketEndpointInputConfig](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/interfaces/bucketendpointinputconfig.html):
> 
> > Whether to force path-style URLs for S3 objects (e.g., `https://s3.amazonaws.com//` instead of `https://.s3.amazonaws.com/`)

Reviewer: ```suggestion
For example:

| `s3PathStyle` | Path                               |
| ------------- | ---------------------------------- |
| `off`         | `https://bucket.s3.amazonaws.com/` |
| `on`          | `https://s3.amazonaws.com/bucket/` |

Read the [AWS S3 docs, Interface BucketEndpointInputConfig](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/interfaces/bucketendpointinputconfig.html) to learn more about path-style URLs.
```

I like @viceice's suggestion above. But quotes must match the actual quote.

How about this instead?

- Use a table to show the difference between config option on/off
- Drop the AWS docs quote
- Rewrite sentence to mention the AWS docs

My suggestion only gives you the _layout_. Please update the table so it shows the _correct_ information!

| `requireConfig=optional` | An onboarding PR will be created if no config file exists. If the onboarding PR is closed and there's no config file, the repository will be processed. | Repository is processed regardless of config file presence.   |
| `requireConfig=ignored`  | No onboarding PR will be created and repo will be processed while ignoring any config file present.                                                     | Repository is processed, any config file is ignored.          |

## s3Endpoint

If set, Renovate will use this string as the `endpoint` when creating the AWS S3 client instance.

## s3PathStyle

If set, Renovate will enable `forcePathStyle` when creating the AWS S3 client instance.

> Whether to force path-style URLs for S3 objects (e.g., `https://s3.amazonaws.com//` instead of `https://.s3.amazonaws.com/`)

Source: [AWS S3 documentation - Interface BucketEndpointInputConfig](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/interfaces/bucketendpointinputconfig.html)

> 

Reviewer 2: @Gabriel-Ladzaretti why is this necessary? Isn't it the same administrator who's configuring the S3?

Reviewer 4: This is mainly used for s3 compatible solutions that dont support or are not configured to use `virtual-hosted–style` requests. It essentially alters the way the aws client constructs and resolves urls.
[path-style-access](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html#path-style-access) 
e.g, `https://s3.amazonaws.com/mybucket/myfile`
vs
[virtual-hosted-style-access](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html#virtual-hosted-style-access) 
e,g, `https://mybucket.s3.amazonaws.com/myfile`

i dont think there is another way to support these s3 instances.

Reviewer 3: I use it for my Minio instance
